test/docker-e2e/e2e_upgrade_test.go

```diff
@@ -0,0 +1,317 @@
+package docker_e2e
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"regexp"
+	"strconv"
+	"strings"
+	"testing"
+
+	"celestiaorg/celestia-app/test/docker-e2e/dockerchain"
+
+	"github.com/celestiaorg/tastora/framework/testutil/wait"
+	tastoratypes "github.com/celestiaorg/tastora/framework/types"
+	"github.com/cosmos/cosmos-sdk/crypto/keyring"
+)
+
+// TallyResponse represents the JSON response from the signal tally query
+// We had to define it again because the JSON response type is not a number but string
+type TallyResponse struct {
+	VotingPower      string `json:"voting_power"`
+	ThresholdPower   string `json:"threshold_power"`
+	TotalVotingPower string `json:"total_voting_power"`
+}
+
+// UpgradeConfig holds configuration parameters for performing an upgrade test
+// on the Celestia chain.
+type UpgradeConfig struct {
+	// BaseBinaryVersion is the initial version of the chain before upgrade.
+	BaseBinaryVersion string
+	// TargetBinaryVersion is the version to which the chain will be upgraded.
+	TargetBinaryVersion string
+	// TargetAppVersion is the app version to upgrade to (used in major upgrades).
+	TargetAppVersion uint64
+}
+
+// TestCelestiaAppMinorUpgrade tests a simple upgrade from one minor version to another.
+func (s *CelestiaTestSuite) TestCelestiaAppMinorUpgrade() {
+	if testing.Short() {
+		s.T().Skip("skipping celestia-app minor upgrade test in short mode")
+	}
+
+	s.RunMinorUpgradeTest(UpgradeConfig{
+		BaseBinaryVersion:   "v4.0.2-mocha",
+		TargetBinaryVersion: "v4.0.7-mocha",
+	})
+}
+
+// TestCelestiaAppMajorUpgrade tests a major upgrade from v4.0.7-mocha to a commit hash which has v5
+// using the signaling mechanism.
+func (s *CelestiaTestSuite) TestCelestiaAppMajorUpgrade() {
+	if testing.Short() {
+		s.T().Skip("skipping celestia-app major upgrade test in short mode")
+	}
+
+	s.RunMajorUpgradeTest(UpgradeConfig{
+		BaseBinaryVersion:   "v4.0.10",
+		TargetBinaryVersion: dockerchain.GetCelestiaTag(),
```

Comment:
[blocking] this test doesn't need to start a v4.0.10 node. A v5.x.x multiplexer node should be able to run a chain that starts on v4 and upgrades to v5.

Reply:
The code is updated to reflect that.